### DISCUSSION TITLE
線上儲存方案的選擇

### DISCUSSION THREAD

**Discussion by gasolin** [lun-dao/LunDAO#31] [General]
雖然最常聽到使用的是 https://ipfs.io/
但 https://www.arweave.org/ 據說能永久儲存，似乎也不失為一個選項

FYR:
去年有發生 Matters 的 IPFS 伺服器掛掉，導致無法訪問 Matters 文章的事件
https://matters.news/@catding/matters%E7%9B%AE%E5%89%8D%E6%98%AF%E4%BD%BF%E7%94%A8-%E5%8E%BB%E4%B8%AD%E5%BF%83%E5%8C%96%E6%8A%80%E8%A1%93-%E5%9C%A8-%E4%B8%AD%E5%BF%83%E5%8C%96%E5%84%B2%E5%AD%98-%E6%96%87%E7%AB%A0-bafyreid5plqwb5xhhzl2ngenq75xcb2nyoar7atcrpke4tt7herje2xwna

**Comment by gasolin** [comment]
另外也有一個選擇是不要做鏈上儲存, 
這樣避免了高額手續費, 

低手續費 -> 可以修正/可以接受初級文章 -> 多一點寫手 -> 大家用 github fork / 儲存在自己電腦中

也可以算是種去中心化

**Comment by yurenju** [comment]
這個我沒研究，但丟到 IPFS 我的理解應該可以不用錢，只是不保證會幫你永續儲存。我想有 Github + IPFS 應該還算是足夠的，畢竟每個人的電腦都可以有一份備份。

**Reply by gasolin** [reply]
恩, 透過 [pinta](https://www.pinata.cloud/) 好像一定量之前都免費 👍

**Comment by wiasliaw** [comment]
IPFS 麻煩點在於沒有被其他的節點 cache 或 pin 會掉資料。以內容作為計算雜湊的依據，所以不便於做文章修改，且修改是需要重新丟 IPFS。有 IPNS 可以做到不更新雜湊就能得新的內容，但印象中需要自己架設節點（可能需要節點的資訊，沒辦法每篇文章都有個 ipns），而且新的內容還是需要重丟。

自己是使用 [nft.storage](https://nft.storage) 和 [web3.storage](https://web3.storage)，除了踩過 SDK 的雷還有上傳時間有點久之外，沒什麼大問題，好處則是完全免費。

如果費用合理，我還是會投 arweave 一票，最近我想將敝司內的 ipfs 都換成 arweave XDD

**Reply by yurenju** [reply]
我想應該可以先從免費的開始，因為 github pages 上面還會有一份，有 clone 過的大家電腦裡應該都還會有一份可以從某個時間點重建，IPFS 作為一個備援的解決方案目前應該是還行的。

另外因為我沒用過 IPFS，想問一下我的想像是不是正確的：
1. static web publish 後可以拿到一個不會改變的 hash 入口點 (CID?)
2. 可以用 ENS 指到這個 CID
3. https://lundao.eth.link 可以連到這個 IPFS 網站
4. 有新內容時，不需要一直修改 ens 所指到的 CID

我這樣的理解對嗎？另外要麻煩你發個 PR 把自己的 ethereum address 加入 [contributors.csv][1]！

[1]: https://github.com/lun-dao/LunDAO/blob/main/contributors.csv

**Reply by wiasliaw** [reply]
1. IPFS 是的，不過整個 static web 丟上去我沒試過
2. \3. 透過 ENS set record 達成，並加上 .link 做到
4. 應該是要將 IPNS 的值設定到 ENS content record，之後網站還是需要重丟並更新到那個 IPNS 下

**Reply by a2468834** [reply]
(1) 其實是有產品實做 IPFS 的 incentive 層，可以考慮搭配用 filecoin
(2) @wiasliaw 說的沒錯，目前最大的問題 (或造成開發者體驗不好) 的地方就是在 garbage collection 機制

IPFS的官網[文章](https://docs.ipfs.io/concepts/persistence/#persistence-versus-permanence)提到：

> To ensure that data persists on IPFS, and is not deleted during garbage collection, data can be pinned to one or more IPFS nodes.
> 為了確保資料能夠持續存在於 IPFS 網路、不受垃圾回收的影響而被刪除，資料必須被釘選（pinned）在至少一個 IPFS 節點上

因此核心的關鍵點就是「如何讓被釘選的資料，在有限時間內被最多的 ipfs 節點探索並暫存起來」
註：gc 預設是一小時，當然可手動調長或短，請參考[這邊](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md)

這邊需要澄清一個常被誤會的想法：
執行 `ipfs add` 之後，我的檔案就永久儲存在 IPFS 網路上了～

這個觀點對也不對。（以下用「A 檔案」來代稱欲 `ipfs add` 的檔案們）

IPFS 是 P2P 網路，存在取決於是否能被其他 IPFS 探索到。
倘若 A 檔案尚不及被其他節點探索之前，你的節點就先下線，那麼等於 A 檔案不曾存在於這個網路；倘若 A 檔案已被其他節點探索過，但是這群節點在整個網路拓撲上面屬於鄰近節點，那麼很高的機率令 A 檔案無法被大部分使用者找到；倘若 A 檔案已被眾多節點探索過，但是你的節點已下線許久，且該檔案久未被其他使用者存取，那麼很高的機率會讓其他節點取消釘選（un-pinned），並隨 gc 而隨風飛逝

---

由於其他 ipfs 節點並未有動機 (incentive) 去主動探詢其他節點的檔案，因此很常遇到自己的檔案遲遲無法被其他人探索，尤其單一檔案大小超過 16MB 以上（個人小範圍實測的數據，非官方所述或嚴謹實驗所得）

先前曾和朋友解決圖片 (metadata.json) 利用 `ipfs add` 指令進行上傳並釘選的過程，需經過很久後其他 IPFS 節點才會發現此檔案（亦即其他節點執行 `ipfs get` 會遲遲無結果）；在這個議題上我們花了不少時間做各種奇奇怪怪操作，最後仍無法找到簡潔、優雅且快速的解決方案。

依循前述的邏輯，我們僅得以下兩類策略（皆非 100% 有效根治問題）

(1) 企圖創造欲散播的檔案具有熱門存取的特性
利用不同的 ip 位址（可搭配 VPN 使用），在許多 VM 裡面皆執行 `ipfs get` 目標檔案的指令；可以搭配使用傳統 web2 方式的 https request 請求 ipfs gateway，效果更佳（ipfs.io 或 cloudflare-ipfs.com 速度都不錯）；定期需要做這項工作，確保檔案隨時可在不同的 ipfs node 被探索到

(2) 自架 ipfs-cluster：請參考此處 https://github.com/ipfs/ipfs-cluster

綜合以上所述，熱愛尋找非中心化服務的我，仍然難以擺脫 Pinata 等的 SaaS，著實是個令人不開心的消息

p.s.1: web3.storage 是被 IPFS 官網推荐的服務之一耶，改天再來研究一下
p.s.2: 果然有沒有 incentive 就是差滿多的 QQQQ

**Comment by a2468834** [comment]
關於 Arweave 訴求比 Filecoin 更安全（資料存在於網路的 durability 更高）且便宜，個人淺見是這樣：

(1) 更安全
個人不學無術，尚未精通博弈論
難以理解 Arweave 利用博弈論來促使所有節點踴躍儲存更多資料
（似乎只有 incentive 而沒有 penalty，不確定如何防止惡意行為）

(2) 更便宜
嗯，Arweave 幣價漲起來就沒有了
~~就如同有些訴求 tx fee 很低的鏈一樣~~
